enterprise/backend/src/metabase_enterprise/workspaces/dependencies.clj

```diff
@@ -0,0 +1,308 @@
+(ns metabase-enterprise.workspaces.dependencies
+  "Workspace-local dependency tracking.
+
+   Unlike the global dependency table which links to table_id (requiring the table to exist and be synced),
+   workspace dependencies link to workspace_input which stores logical table references that don't require
+   the table to exist yet."
+  (:require
+   [clojure.set :as set]
+   [metabase-enterprise.workspaces.models.workspace-dependency]
+   [metabase-enterprise.workspaces.models.workspace-input]
+   [metabase-enterprise.workspaces.models.workspace-output]
+   [metabase-enterprise.workspaces.util :as ws.u]
+   [metabase.driver :as driver]
+   [metabase.driver.sql :as driver.sql]
+   [metabase.lib.core :as lib]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]
+   [toucan2.core :as t2]))
+
+;;; ---------------------------------------- Schemas ----------------------------------------
+
+(mr/def ::table-ref
+  "A logical reference to a table by schema and name. Includes table_id if the table exists."
+  [:map
+   [:database_id :int]
+   [:schema [:maybe :string]]
+   [:table :string]
+   [:table_id {:optional true} [:maybe :int]]])
+
+(mr/def ::output-ref
+  "A reference to an output table (transform target)."
+  [:map
+   [:database_id :int]
+   [:schema [:maybe :string]]
+   [:table :string]])
+
+(mr/def ::analysis
+  "The result of analyzing an entity's dependencies."
+  [:map
+   [:output ::output-ref]
+   [:inputs [:sequential ::table-ref]]])
+
+;;; ---------------------------------------- Batch table lookups ----------------------------------------
+
+(defn- batch-table-refs-from-ids
+  "Look up table metadata for multiple table IDs in a single query.
+   Returns a map of table-id -> {:database_id :schema :table :table_id}."
+  [table-ids]
+  (when (seq table-ids)
+    (->> (t2/select [:model/Table :id :db_id :schema :name] :id [:in table-ids])
+         (into {} (map (fn [{:keys [id db_id schema name]}]
+                         [id {:database_id db_id
+                              :schema      schema
+                              :table       name
+                              :table_id    id}]))))))
+
+(defn- batch-lookup-table-ids
+  "Look up table_ids for multiple logical table references in a single query.
+   Uses the same matching logic as find-table-or-transform: if parsed schema is nil,
+   it matches against the driver's default schema.
+   Returns a map of [database_id search-schema table] -> table_id where search-schema
+   is the schema used for lookup (defaulted if nil)."
+  [driver database-id table-refs]
+  (when (seq table-refs)
+    (let [default-schema (driver.sql/default-schema driver)
+          tables         (t2/select [:model/Table :id :schema :name]
+                                    :db_id database-id
+                                    :name [:in (map :table table-refs)])
```

Comment:
Concern: I don't like the over-fetch - imagine if one table exists in 1000 schemas.

This isn't actually that far fetched, if people are using metabot codegen like crazy, and its generating lots of workspaces.

At the very least using `:in` on schema as well would mitigate things a lot, even though there would still be over-fetch. That over-fetch might still be better than doing `[:or [:and ...] ...]`, and it's much easier.
```suggestion
          tables         (t2/select [:model/Table :id :schema :name]
                                    :db_id database-id
                                    :schema [:in (map :schema table-refs)]
                                    :name [:in (map :table table-refs)])
```